### PR TITLE
feat(DefinitionTooltip): add Definition Tooltip

### LINKED ISSUE #6429: TootlipDefinition: Support other values than 'strings' in the children prop
### Summary

Change `TooltipDefinition` React component to support other value types of the `children` prop than `string`. 

### Justification

So we can show the TooltipDefinition on other UI elements like `h2`, `<p>` without any console warnings.

image

```jsx
<TooltipDefintion tooltipText="Banana">
  <h3>Hello</h3>
</TooltipDefintion>;
```

Unfortunately wrapping the `TooltipDefintion` definition in another component won't work as the tooltip is rendered as a button:

```jsx
<h3> // Doesn't work
  <TooltipDefintion tooltipText="Banana">Hello</TooltipDefintion>;
</h3>;

```

### Available extra resources

I can help make this change if it's accepted.

The PropType below will need to be changed to `PropTypes.node`

https://github.com/carbon-design-system/carbon/blob/36574d8cc68e46f00c7b52f230618eb1b1e3a131/packages/react/src/components/TooltipDefinition/TooltipDefinition.js#L97

### LINKED ISSUE #8814: Implement definition tooltip styles for Tooltip next
We currently have our base implementation of the next `Tooltip` component, this issue is to expand this implementation to include the styles and React component for the Definition tooltip variant.

Link: https://www.carbondesignsystem.com/components/tooltip/usage/#definition-tooltip

### LINKED ISSUE #9902: Determine implementation for definition tooltip
[Usage docs](https://www.carbondesignsystem.com/components/tooltip/usage/#definition-tooltip)

This component replaces `TooltipDefinition`. It is used in the flow of text in order to provide more details about a term. The challenge with this component is that there isn't a clear interactive element to be used for the trigger. Currently, we use a `button` where the label of the button is the term and the description of the button is the contents of the tooltip. However, this means that activating the button does nothing and could be confusing to the end-user.

This issue is to determine how we want to implement this component in v11. It should document the HTML structure, interactions, and anything else that will be used next sprint to build out this component in React.

**Behavior**

- Appear on hover, disappear on mouseout
- Focus, press enter or space, appears and is persistent
- Focused, visible, press enter or space, disappears
- Focused, visible, press esc to dismiss?

**Questions**

- Does the definition tooltip contain interactive descendants? Yes, it could
- ~~Is this something that `role="term"` and `role="definition"` could help us out with?~~

### PR DESCRIPTION
**Pull request by dakahn**
Closes #9902
Closes #8814
Closes #6429
Refs #9901

(merged the last PR https://github.com/carbon-design-system/carbon/pull/10197 on accident too soon. So reverted that and HERE WE GO AGAIN)

This PR adds the DefinitionTooltip component (utilizing our Popover primitive 🍖) and DefinitionTooltip styles ⚡ 
image

#### Testing / Reviewing

In the `carbon-react` package run the storybook and check out the unstable_tooltip > Definition story. Make sure it matches intended spec and behavior. 🏄🏾

### CODE REVIEW

File: packages/react/src/components/Tooltip/next/__tests__/Tooltip-test.js [hunk 1]
```diff
@@ -55,3 +57,42 @@ describe('Tooltip', () => {
     expect(screen.getByText('test')).toHaveAttribute('aria-describedby');
   });
 });
+
+describe('DefintiionTooltip', () => {
```
Comment (joshblack) [inline packages/react/src/components/Tooltip/next/__tests__/Tooltip-test.js]: I think we should split this out into its own test file at `next/__tests__/DefinitionTooltip-test.js` to make it easier to find 👍

**Review by joshblack — APPROVED** [review]
Looks good, only comment would be to move tests over into its own test file 👍

File: packages/react/src/components/Tooltip/next/__tests__/Tooltip-test.js (hunk 1 above)
Comment (dakahn) [reply packages/react/src/components/Tooltip/next/__tests__/Tooltip-test.js]: On it

**Review by tay1orjones — APPROVED** [review]

**Review by laurenmrice — CHANGES_REQUESTED** [review]
@dakahn 

**Hover:**
- [x] The dotted underline should be color token `$border-interactive`.
Artboardblue

**Popover:**
~~Caret tip should be 16px to the right.~~ This will be tackled in a separate PR for Popover. Definition tooltip will inherit the change.
Artboard

**Safari bug:**
- [x] When clicking on the trigger text, the text disappears briefly and there is no focus box.
safari click bug

**Theme switching:**
When switching themes, there a problem with the theme switcher It seems like the `p` text is using the correct color token `$text-primary`, but it is visually not showing up correctly when I change themes. The `URL` trigger text also appears to be the wrong color.
Screen Shot 2021-12-09 at 10 32 41 AM

**On click option:**
- [ ]  Can we add a prop to have the option to make the tooltip appear `on click` instead of on hover? We have heard some teams prefer to have the definition tooltip appear on click because of their product usecases. We would want to keep the hovering interaction as our default still, but would like to add an option for to open the tooltip `on click`.

**Comment by joshblack** [comment]
@laurenmrice for this point:

>  Caret tip should be 16px to the right.

Is this something specific for definition tooltip? I believe the reason that it's centered is that we made that update to popover so that the caret is placed center of the target so just wanted to confirm.

> When switching themes, is there a problem with the theme switcher?

I think this should be fixed in latest, I believe we didn't have `color: $text-primary` on the `body` for the default color 👍

**Comment by laurenmrice** [comment]
@joshblack I just double checked the PR of the popover refactor and I missed in that visual review that the caret tip is not aligned to the 16px padding. It should be aligned to the type that is following the 16px padding and the caret tip should still be vertically center with the trigger button. So maybe this is something that needs to be fixed for Popover and then inherited for Definition tooltip?

Example alignment:
Artboard 1

**Comment by joshblack** [comment]
> So maybe this is something that needs to be fixed for Popover and then inherited for Definition tooltip?

Definitely, that makes sense to me 👍

**Review by laurenmrice — CHANGES_REQUESTED** [review]
@dakahn 

Enabled (closed) state
- [x] The dotted line needs to use color token $border-strong for the Enabled (closed) state. On Hover and Active states it changes to color token $border-interactive.
Artboard

Theming
- [x] The text color of the definition tooltip in the sentence is showing up as the wrong color in the dark themes. The text color should be using color token $text-primary.

 Example:
Screen Shot 2022-01-18 at 10 58 34 AM


My previous comment about having an “on click option”

> Can we add a prop to have the option to make the tooltip appear on click instead of on hover? We have heard some teams prefer to have the definition tooltip appear on click because of their product usecases. We would want to keep the hovering interaction as our default still, but would like to add an option for to open the tooltip on click.

Is this going to be tackled in a separate PR?

**Comment by dakahn** [comment]
@laurenmrice i should be able to get that prop in. Thanks for the review 🏄🏾

**Comment by dakahn** [comment]
@laurenmrice styling fixes should be in -- api additions with the 'on click' on the way

**Comment by dakahn** [comment]
@laurenmrice added the openOnClick prop. I've updated the storybook story to use it by default for testing 👍🏾

**Review by laurenmrice — CHANGES_REQUESTED** [review]
**openOnClick prop:**
- [x] Even though you have to click on the text to open the tooltip, there should still be a hover color change of the underline. The color token on hover should be $border-interactive.
hover color


**Safari bug:**
- [ ] Once the definition tooltip is open, you can not close the tooltip by clicking anywhere on the page and you should be able to.  

tooltip close
----

- Everything else looks good! 👍

**Comment by dakahn** [comment]
@laurenmrice I can recreate the Safari bug, but I'm not sure whats causing it. For other dev reviewers, it look like state is being set properly and onBlur is being called as expected 😕

**Comment by tw15egan** [comment]
In my past experience, sometimes the `onBlur` doesn't trigger in Safari since button clicks technically don't give the element focus in Safari. I think you may need to manually focus the element in the `onClick`

```js
onClick={(evt) => {
    evt.target.focus();
    setOpen(!isOpen);
}}
```

I can't test this, because mine works in Safari for some reason 🤷🏻‍♂️

**Review by laurenmrice — APPROVED** [review]
Looks good to me now! Thanks D.A.! 🎉

**Comment by joshblack** [comment]
Anything left for this one @dakahn? :eyes:

**Comment by dakahn** [comment]
@joshblack  yep! just a couple broken tests im patchin up rn. should be ready to go in the next hour tho 👍🏾

**Pull request merged**